pandas/core/dtypes/astype.py

```diff
@@ -246,3 +247,39 @@ def astype_array_safe(
             raise
 
     return new_values
+
+
+def astype_is_view(dtype: DtypeObj, new_dtype: DtypeObj) -> bool:
+    """Checks if astype avoided copying the data.
+
+    Parameters
+    ----------
+    dtype : Original dtype
+    new_dtype : target dtype
+
+    Returns
+    -------
+    True if new data is a view, False otherwise
+    """
+    if dtype == new_dtype:
+        return True
+
+    elif isinstance(dtype, np.dtype) and isinstance(new_dtype, np.dtype):
+        # Only equal numpy dtypes avoid a copy
+        return False
+
+    if is_string_dtype(dtype) and is_string_dtype(new_dtype):
+        return True
```

Comment:
I would think this is not _always_ guaranteed to be a view? (but of course safer to return True than incorrectly assume it is always a copy) 

`is_string_dtype` also returns true for generic "object", and converting object to string is not necessarily no-copy

Reply:
Using this branch, I see the following wrong behaviour in case you have mixed objects:

```
In [1]: s = pd.Series(['a', 'b', 1])

In [2]: s2 = s.astype("string")

In [3]: s.values
Out[3]: array(['a', 'b', 1], dtype=object)

In [4]: pd.options.mode.copy_on_write = True

In [5]: s = pd.Series(['a', 'b', 1])

In [6]: s2 = s.astype("string")

In [7]: s.values
Out[7]: array(['a', 'b', '1'], dtype=object)
```

Because of not taking a copy, the `ensure_string_array` actually mutates the original values in place.

For the case of "object -> some extension dtype" casting, we should probably always do `copy=True`, because I don't think we can rely on `_from_sequence` to do the correct thing (although for this specific case I also don't think that `StringArray._from_sequence(arr, copy=False)` should mutate `arr` in place. I would expect to _only_ not make a copy if it's not needed, i.e. if it can just view the original data and no modification needs to be done)

Reply:
I opened https://github.com/pandas-dev/pandas/pull/51073 to address the inplace modification.

You are correct, it is not guaranteed to be a no-copy op, but it could be and I couldn't figure out a more precise check.

Reply:
> We can still optimise in a follow-up to get this stricter, for now I was aiming to get as many cases as possible without making it overly complex

Can you add a comment explaining the True and  to note something like that?